worker/src/direct_invocation/ws_handler.rs

```diff
@@ -0,0 +1,253 @@
+use codec::{Decode, Encode};
+use log::*;
+use sp_core::H256 as Hash;
+use std::sync::Arc;
+use substratee_enclave_api::EnclaveApi;
+use substratee_worker_primitives::{DirectRequestStatus, RpcResponse, RpcReturnValue};
+use ws::{CloseCode, Handler, Message, Result, Sender};
+
+/// Trait for an abstract factory to create a handler instance
+pub trait CreateWsHandler {
+    type Handler: Handler;
+
+    /// create a WS handler based on a sender
+    fn create(&self, sender: Sender) -> Self::Handler;
+}
+
+/// WS handler factory implementation, requires the enclave ID and a watch_list facade
+pub struct WsHandlerFactory<E, W> {
+    enclave_api: E,
+    watch_list: Arc<W>,
+}
+
+impl<E, W> WsHandlerFactory<E, W> {
+    pub fn new(enclave_api: E, watch_list: Arc<W>) -> Self {
+        WsHandlerFactory {
+            enclave_api,
+            watch_list,
+        }
+    }
+}
+
+impl<E, W> CreateWsHandler for WsHandlerFactory<E, W>
+where
+    E: EnclaveApi + Clone,
+    W: WatchList<<WsSender as CreateWatchingClient>::Client>,
+{
+    type Handler = WsHandler<WsSender, E, W>;
+
+    fn create(&self, sender: Sender) -> Self::Handler {
+        WsHandler {
+            sender: WsSender::new(sender),
+            enclave_api: self.enclave_api.clone(),
+            watch_list: self.watch_list.clone(),
+        }
+    }
+}
+
+/// WebSocket Handler implementation
+pub struct WsHandler<Sender, E, W> {
+    sender: Sender,
+    enclave_api: E,
+    watch_list: Arc<W>,
+}
+
+impl<Sender, E, W> Handler for WsHandler<Sender, E, W>
+where
+    Sender: WsSend + CreateWatchingClient,
+    E: EnclaveApi,
+    W: WatchList<Sender::Client>,
+{
+    fn on_message(&mut self, msg: Message) -> Result<()> {
+        if self
+            .handle_direct_invocation_request(msg.to_string().as_str())
+            .is_err()
+        {
+            error!("direct invocation call was not successful");
+        }
+        Ok(())
+    }
+
+    fn on_close(&mut self, code: CloseCode, reason: &str) {
+        debug!(
+            "Direct invocation WebSocket closing for ({:?}) {}",
+            code, reason
+        );
+    }
+}
+
+impl<Sender, E, W> WsHandler<Sender, E, W>
+where
+    Sender: WsSend + CreateWatchingClient,
+    E: EnclaveApi,
+    W: WatchList<Sender::Client>,
+{
+    fn handle_direct_invocation_request(&self, request_message: &str) -> Result<()> {
+        info!("Got message '{}'. ", request_message);
+
+        let msg: Vec<u8> = request_message.as_bytes().to_vec();
+
+        let response = self
+            .enclave_api
+            .rpc(msg)
+            .map_err(|e| ws::Error::new(ws::ErrorKind::Internal, format!("{:?}", e)))?;
+
+        let decoded_response = String::from_utf8_lossy(&response).to_string();
+
+        let full_rpc_response = (serde_json::from_str(&decoded_response)
+            as serde_json::Result<RpcResponse>)
+            .map_err(|e| ws::Error::new(ws::ErrorKind::Internal, format!("{:?}", e)))?;
```

Comment:
The below works for me. What did the compiler tell you? :)
```
let full_rpc_response = serde_json::from_str(&decoded_response)
            .map_err(|e| ws::Error::new(ws::ErrorKind::Internal, format!("{:?}", e)))?;
```

Reply:
But we can redude L114-L118 to the following:
```
let full_rpc_response = serde_json::from_slice(&response)
     .map_err(|e| ws::Error::new(ws::ErrorKind::Internal, format!("{:?}", e)))?;
```

Reply:
Yes the cast can be removed, compiler thinks it's fine 👍  This was just a legacy from before I guess